neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,164 @@
+---
+NEP: 519
+Title: Yield Execution
+Authors: Akhi Singhania <akhi3030@gmail.com>
+Status: Draft
+DiscussionsTo: https://github.com/near/NEPs/pull/519
+Type: Protocol
+Version: 0.0.0
+Created: 2023-11-17
+LastUpdated: 2023-11-20
+---
+
+## Summary
+
+Today, when a smart contract is called by a user or another contract, it has no sensible way to delay responding to the caller.  There exist use cases where contracts would benefit from being able to delay responding till some arbitrary time in the future.  This proposal introduces such possibility into the NEAR protocol.
+
+## Motivation
+
+There exist some situations where when a smart contract on NEAR is called, it will only be able to provide an answer at some arbitrary time in the future.  So the callee needs a way to defer replying to the caller till this time in future.
+
+Examples include when a smart contract (`S`) provides MPC signing capabilities parties external to the NEAR protocol are computing the signature.  The rough steps are:
+1. Signer contract provides a function `fn sign_payload(Payload, ...)`.
+2. When called, the contract updates some contract state which is being monitored by external indexers to indicate that a new signing request has been received.  It also defers replying to the caller.
+3. The indexers observe the new signing request, they compute a signature and call another function `fn signature_available(Signature, ...)` on the signer contract.
+4. The signer contract validates the signature and if validate, replies to the original caller.
+
+Today, the NEAR protocol has no sensible way to defer replying to the caller in step 2 above.  This proposal proposes adding two following new host functions to the NEAR protocol:
+
+- `yield`: this can be called by a contract to indicate to the protocol that it is not ready yet to reply to its caller.
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
+
+## Specification
+
+The proposal is to add the following host functions to the NEAR protocol:
+
+
+```rust
+/// Instructs the protocol that the smart contract is not ready yet to respond
+/// to its caller yet.  The smart contract promises to call `yield_resume()`
+/// within `yield_num_blocks` blocks.  When `yield_resume()` is called, the
+/// protocol will call the method on the smart contract that is identified by
+/// `method_name_len` and `method_name_ptr` and this method will be expected to
+/// either respond to the caller or create another promise.
```

Comment:
```suggestion
/// `method_name_len` and `method_name_ptr` and this method may respond to the caller.
/// Once the method has responded, `yield_resume` can no longer be called on this promise.
```
The difference here is that the promise index remains the same with multiple calls to resume. The current design means that you need to have at most one resume call in flight at a time or the second one will fail.

Reply:
Ah interesting, yes I think we should support this.  I will add this suggestion.  I don't have a good intuition for whether or not it will make the implementation trickier.  Let's find out.